src/librustc_mir/transform/const_prop.rs

```diff
@@ -257,13 +257,21 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for ConstPropMachine {
         l.access()
     }
 
-    fn before_access_static(
+    fn before_access_global(
         _memory_extra: &(),
+        _alloc_id: AllocId,
         allocation: &Allocation<Self::PointerTag, Self::AllocExtra>,
+        def_id: Option<DefId>,
+        is_write: bool,
     ) -> InterpResult<'tcx> {
-        // if the static allocation is mutable or if it has relocations (it may be legal to mutate
-        // the memory behind that in the future), then we can't const prop it
-        if allocation.mutability == Mutability::Mut || allocation.relocations().len() > 0 {
+        if is_write {
+            throw_unsup!(ConstPropUnsupported("can't write to global"));
+        }
+        // If the static allocation is mutable or if it has relocations (it may be legal to mutate
+        // the memory behind that in the future), then we can't const prop it.
+        // FIXME: we only check statics here (that have a `DefId`), not other mutable allocations.
+        // Why that?
+        if def_id.is_some() && (allocation.mutability == Mutability::Mut || allocation.relocations().len() > 0) {
```

Comment:
This strikes me as potentially dangerously incorrect: the new check matches the old one, which was only called for `Static` global allocations, not for `Mem`. But that means we still const-prop from mutable global `Mem` allocations! Couldn't that be incorrect, as the value might be different at run-time? Cc @wesleywiser

Reply:
yea... not right now, but in the future this will become a problem. Right now we can't end up with mutable persistent memory other than mutable statics (mutable constants don't count, as they are not "persistent" in the sense that reading from them will always yield the same value, even if they have interior mutability)

Reply:
I don't follow, what do you mean by "persistent"?

The code here is definitely reachable for mutable non-static globals -- and to my great irritation, this happens during *validation*. I am not sure why const-prop performs validation?

Reply:
there are no mutable non-static globals afaik? I think what you are encountering is Allocations with their mutability bit set, but which are never used in a way that creates pointers to them.

Reply:
> I think what you are encountering is Allocations with their mutability bit set

"mutability bit set" means they are mutable, doesn't it?

> but which are never used in a way that creates pointers to them.

What does that mean?

Reply:
Consider a `const FOO: Cell<BigThingNotScalar> = Cell::new(BigThingNotScalar);`. You get a mutable allocation because of the `Cell`, but you can never mutate that allocation.

Reply:
Hm, I see. If that is so, is there a way we can make these allocations immutable? No reason to put them into read-write memory.

Reply:
should be no problem. Interning is already aware of what it is interning. I have a feeling of déja vu though.

Reply:
The comment here claims that we already intern such things as immutable:

https://github.com/rust-lang/rust/blob/342c5f33d097b2dc07a2dbc0ca45a37379d2ff60/src/librustc_mir/interpret/intern.rs#L122-L125

So either this does not work, or the issue here is a different one.
Anyway the issue is pre-existing, not introduced by this PR, so shall I just open an issue to track it?

Reply:
yea, that seems the best

Reply:
Okay turns out the problem wasn't mutability, it was that there are relocations in there. I expanded the mutability check to cover all globals now, and tests still pass locally.